src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -85,10 +85,11 @@ internal enum TerminatorState
             IsEndOfFunctionPointerParameterListErrored = 1 << 24,
             IsEndOfFunctionPointerCallingConvention = 1 << 25,
             IsEndOfRecordOrClassOrStructOrInterfaceSignature = 1 << 26,
-            IsExpressionOrPatternInLabelOfSwitchStatementOrExpression = 1 << 27,
+            IsExpressionOrPatternInCaseLabelOfSwitchStatement = 1 << 27,
+            IsPatternInSwitchExpressionArm = 1 << 28,
```

Comment:
Did you consider using the same TerminatorState flag in both scenarios and using the same recovery strategy (abort on either an unexpected `:` or an unexpected `=>`)?